sql/sql_expression.cc

```diff
@@ -3119,49 +3128,246 @@ bind(SqlBindingScope & context) const
-                     ExpressionValue & storage,
-                     const VariableFilter & filter) -> const ExpressionValue &
-        {
-            StructValue result;
-            result.reserve(boundClauses.size());
-            for (auto & c: boundClauses) {
-                ExpressionValue storage;
-                const ExpressionValue & v = c(context, storage, filter);
-                if (&v == &storage)
-                    storage.mergeToRowDestructive(result);
-                else v.appendToRow(Path(), result);
+        // Have we already done this prefix?
+        std::vector<int> donePrefixes(allPrefixes.size(), 0);
+
+        struct ClauseInstruction {
+            ClauseInstruction(PathElement el = PathElement(),
+                              int outputNumber = -1,
+                              bool isFirst = false)
+                : el(std::move(el)),
+                  outputNumber(outputNumber),
+                  isFirst(isFirst)
+            {
+            }
+
+            PathElement el;
+            int outputNumber = -1;  ///< Where do we put it?
+            bool isFirst = false;   ///< What do we do with it?
+
+            void addToResult(ExpressionValue row,
+                             StructValue & result) const
+            {
+                if (isFirst) {
+                    result[outputNumber] = std::make_tuple(el, std::move(row));
+                }
+                else {
+                    std::get<1>(result[outputNumber])
+                        = ExpressionValue::superpose({std::get<1>(result[outputNumber]),
+                                                      std::move(row)});
+                }
+            }
+        };
+
+        // For each clause, for each prefix, what do we do with it?
+        struct ClauseInstructions {
+            /// Expression for the clause
+            BoundSqlExpression expr;
+
+            /// Instructions for this particular clause
+            std::vector<ClauseInstruction> clauseInstructions;
+
+            ClauseInstructions(BoundSqlExpression expr,
+                               std::vector<ClauseInstruction> clauseInstructions)
+                : expr(std::move(expr)),
+                  clauseInstructions(std::move(clauseInstructions))
+            {
+            }
+
+            void apply(const SqlRowScope & context,
+                       const VariableFilter & filter,
+                       StructValue & result) const
+            {
+                // 1.  Apply the clause and get its value
+
+                ExpressionValue clauseStorage;
+                const ExpressionValue & v = expr(context, clauseStorage, filter);
+
+                StructValue row;
+                if (&v == &clauseStorage) {
+                    clauseStorage.mergeToRowDestructive(row);
+                }
+                else {
+                    v.appendToRow(Path(), row);
+                }
+
+                //cerr << "clause " << clauses[i]->surface
```

Comment:
The logger has a major overhead a) it needs to be passed everywhere, and b) calling it even when disabled increases code size and runtime.  Leaving them commented out lets a future debugger start off (possibly with updates to the comments) where left off.